File: y2018/python/day4.py
from dateutil import parser
from datetime import timedelta
from enum import Enum
from collections import defaultdict
import re


class Action(Enum):
    SLEEP = 1
    WAKE = 2
    SHIFT = 3


logs = []
with open('input/day4.txt') as read:
    for log in read.readlines():
        time = log[log.find('[') + 1:log.find(']')]
        action = log[len(time) + 2:].strip()

        guard_string = r'Guard #(\d+) begins shift'
        time = parser.parse(time)
        if action == 'falls asleep':
            logs.append((time, Action.SLEEP))
        elif action == 'wakes up':
            logs.append((time, Action.WAKE))
        elif re.match(guard_string, action):
            guard_id = int(next(iter(re.findall(guard_string, action))))
            logs.append((time, Action.SHIFT, guard_id))

logs.sort()

shift_started = [None]  # first element for edge case
minutes_asleep = defaultdict(lambda: defaultdict(int))
sleep_time = {}
for log in logs:
    guard = shift_started[-1]
    if log[1] == Action.WAKE:
        start = sleep_time[guard]
        while start < log[0]:
            minutes_asleep[guard][start.minute] += 1
            start += timedelta(minutes=1)
    elif log[1] == Action.SLEEP:
        sleep_time[guard] = log[0]
    elif log[1] == Action.SHIFT:
        shift_started.append(log[2])

worst_guard = -1
most_slept = 0
for g, s in minutes_asleep.items():
    time_asleep = sum(s.values())
    if time_asleep > most_slept:
        worst_guard = g
        most_slept = time_asleep

worst_minute = -1
minute_most = 0
for minute, minute_amt in minutes_asleep[worst_guard].items():
    if minute_amt > minute_most:
        minute_most = minute_amt
        worst_minute = minute

print(f"minute slept on most for worst guard: {worst_guard * worst_minute}")

# guard, minute, and how long that minute was slept on
abs_worst_minute = (-1, -1, 0)
for g, s in minutes_asleep.items():
    for minute, minute_amt in s.items():
        if minute_amt > abs_worst_minute[2]:
            abs_worst_minute = (g, minute, minute_amt)

print(f"absolute worst minute: {abs_worst_minute[0] * abs_worst_minute[1]}")

File: y2018/rust/src/bin/day4.rs
use std::collections::HashMap;
use std::fs;

use chrono::{Duration, NaiveDateTime, Timelike};
use regex::Regex;

#[derive(Debug)]
enum Event {
    Sleep,
    Wake,
    Begin(u32),
}

fn main() {
    let event_fmt = Regex::new(r"\[(.*)] (.*)").unwrap();
    let guard_fmt = Regex::new(r"guard #(\d+) begins shift").unwrap();
    let mut events: Vec<(NaiveDateTime, Event)> = fs::read_to_string("../input/day4.txt")
        .unwrap()
        .lines()
        .map(|e| {
            let raw = event_fmt.captures(e).unwrap();
            let time = NaiveDateTime::parse_from_str(&raw[1], "%Y-%m-%d %H:%M").unwrap();
            let what = &raw[2].to_lowercase();
            return (
                time,
                if what == "wakes up" {
                    Event::Wake
                } else if what == "falls asleep" {
                    Event::Sleep
                } else {
                    let guard_id = guard_fmt.captures(what).unwrap()[1].parse::<u32>().unwrap();
                    Event::Begin(guard_id)
                },
            );
        })
        .collect();

    events.sort_by(|a, b| a.0.cmp(&b.0));

    let mut minutes_slept: HashMap<u32, HashMap<u32, u32>> = HashMap::new();
    let mut curr_guard = u32::MAX;
    let mut last_time = None;
    for (t, e) in &events {
        match e {
            Event::Sleep => {
                last_time = Some(*t);
            }
            Event::Wake => {
                let mut curr_at = last_time.unwrap();
                assert_ne!(curr_guard, u32::MAX);
                if !minutes_slept.contains_key(&curr_guard) {
                    minutes_slept.insert(curr_guard, HashMap::new());
                }
                while curr_at < *t {
                    minutes_slept
                        .get_mut(&curr_guard)
                        .unwrap()
                        .entry(curr_at.minute())
                        .and_mshouldodify(|freq| *freq += 1)
                        .or_insert(1);
                    curr_at += Duration::minutes(1);
                }
            }
            Event::Begin(id) => {
                curr_guard = *id;
            }
        }
    }

    let guard_prod =
        |(g, m): (&u32, &HashMap<u32, u32>)| g * m.iter().max_by_key(|i| i.1).unwrap().0;

    let strat1 = minutes_slept
        .iter()
        .max_by(|(_, m1), (_, m2)| m1.values().sum::<u32>().cmp(&m2.values().sum::<u32>()))
        .map(guard_prod)
        .unwrap();

    println!("result of strategy 1: {strat1}");

    let strat2 = minutes_slept
        .iter()
        .max_by(|(_, m1), (_, m2)| m1.values().max().cmp(&m2.values().max()))
        .map(guard_prod)
        .unwrap();

    println!("result of strategy 2: {strat2}");
}
